src/common/particleML.f90

```diff
@@ -68,4 +69,89 @@ module particleML
 !> will be allocated in allocatefieldsml::allocatefields
     type(particle), allocatable, public, save :: pdata(:)
 
+    contains
+      !> Scale activity of the particle,
+      !>
+      !> Activity remaining in the particle is
+      !> factor * original
+      !> The "lost" activity is returned
+      real function scale_rad(p, factor) result(removed)
+        class(particle), intent(inout) :: p
+        real, value :: factor
+        real :: previous
+
+        if (p%is_inactive()) then
+          removed = 0.0
+          return
+        endif
+
+        factor = min(factor, 1.0)
+        factor = max(factor, 0.0)
+
+        previous = p%rad_
+        p%rad_ = factor * previous
+        call p%flush_away_denormal()
+
+        removed = previous - p%rad_
+      end function
+
+      !> Set activity of particle
+      subroutine set_rad(p, rad)
+        class(particle), intent(inout) :: p
+        real, intent(in) :: rad
+        p%rad_ = rad
+      end subroutine
+
+      !> Get activity of particle
+      real pure function get_rad(p)
+        class(particle), intent(in) :: p
+        get_rad = p%rad_
+      end function
+
+      !> Get current activity and set activity
+      real function get_set_rad(p, rad) result(previous)
+        class(particle), intent(inout) :: p
+        real, intent(in) :: rad
+        previous = p%rad_
+        p%rad_ = rad
+      end function
+
+      !> Add activity to the particle
+      subroutine add_rad(p, rad)
+        class(particle), intent(inout) :: p
+        real, intent(in) :: rad
+        p%rad_ = p%rad_ + rad
+      end subroutine
+
+      !> Deactivate a particle
+      !>
+      !> Lost activity is stored as negative
+      subroutine inactivate(p)
+        class(particle), intent(inout) :: p
+        if (p%is_inactive()) return
+        p%rad_ = -p%rad_
+      end subroutine
+
+      !> Get status of particle
+      logical pure function is_inactive(p)
+        class(particle), intent(in) :: p
```

Comment:
Please remove this function! .not. p%is_active() short enough and doesn't require a function of it's own.